bagelbids/tests/test_utility.py

```diff
@@ -0,0 +1,159 @@
+
+
+@pytest.mark.parametrize(
+    "tool, columns",
+    [
+        ("cogAtlas:1234", ["tool_item1", "tool_item2"]),
+        ("cogAtlas:4321", ["other_tool_item1"]),
+    ],
+)
+def test_map_tools_to_columns(load_ex_json, tool, columns):
+    data_dict = load_ex_json("example6.json")
+
+    result = map_tools_to_columns(data_dict)
+
+    assert result[tool] == columns
+
+
+def test_get_transformed_categorical_value(test_data, load_ex_json):
+    """Test that the correct transformed value is returned for a categorical variable"""
+    data_dict = load_ex_json("example2.json")
+    pheno = pd.read_csv(test_data / "example2.tsv", sep="\t")
+
+    assert "bids:Male" == get_transformed_values(
+        columns=["sex"],
+        row=pheno.iloc[0],
+        data_dict=data_dict,
+    )
+
+
+@pytest.mark.parametrize(
+    "value,column,expected",
+    [
+        ("test_value", "test_column", True),
+        ("does not exist", "test_column", False),
+        ("my_value", "empty_column", False),
+    ],
+)
+def test_missing_values(value, column, expected):
+    """Test that missing values are correctly detected"""
+    test_data_dict = {
+        "test_column": {"Annotations": {"MissingValues": ["test_value"]}},
+        "empty_column": {"Annotations": {}},
+    }
+
+    assert is_missing_value(value, column, test_data_dict) is expected
+
+
+@pytest.mark.parametrize(
+    "subject_idx, is_avail",
+    [(0, False), (2, False), (4, True)],
+)
+def test_get_assessment_tool_availability(
+    test_data, load_ex_json, subject_idx, is_avail
+):
+    """
+    Ensure that subjects who have one or more missing values in columns mapped to an assessment
+    tool are correctly identified as not having this assessment tool
+    """
+    data_dict = load_ex_json("example6.json")
+    pheno = pd.read_csv(test_data / "example6.tsv", sep="\t")
+    test_columns = ["tool_item1", "tool_item2"]
+
+    assert (
+        are_not_missing(test_columns, pheno.iloc[subject_idx], data_dict)
+        is is_avail
+    )
+
+
+@pytest.mark.parametrize(
+    "raw_age,expected_age,heuristic",
+    [
+        ("11,0", 11.0, "bg:euro"),
+        ("90+", 90.0, "bg:bounded"),
+        ("20-30", 25.0, "bg:range"),
+        ("20Y6M", 20.5, "bg:iso8601"),
+        ("P20Y6M", 20.5, "bg:iso8601"),
+        ("20Y9M", 20.75, "bg:iso8601"),
+    ],
+)
+def test_age_gets_converted(raw_age, expected_age, heuristic):
+    assert expected_age == transform_age(raw_age, heuristic)
+
+
+def test_invalid_age_heuristic():
+    """Given an age transformation that is not recognized, returns an informative ValueError."""
+    with pytest.raises(ValueError) as e:
+        transform_age("11,0", "bg:birthyear")
+
+    assert "unrecognized age transformation" in str(e.value)
+
+
+# TODO: Probably better to move this function to utility module once it's created, to reuse.
+def _get_models_fields():
+    models_fields_list = []
+    for cname, cobj in inspect.getmembers(models, predicate=inspect.isclass):
```

Comment:
The purpose was to provide a straightforward way to get a list of all the models and fields in `models.py`,  so that the list items could be passed individually as parameters to `test_generate_context()`, which then tests separately that each item exists in the context dictionary outputted by `generate_context()`. 

(I could have iterated through the models/fields directly inside `test_generate_context()` to assert over them one by one, but I didn't like that this produced a single PASS/FAIL. Meanwhile, having individual models/fields as separate test items makes explicit if specific item(s) are missing from the generated `@context`.)

It's true that I could have reused the syntax used inside `generate_context()` in the current function, i.e.
```python
    for val in dir(models):
        klass = getattr(models, val)
        if not isinstance(klass, pydantic.main.ModelMetaclass):
            continue
        fields[klass.__name__] = "bg:" + klass.__name__
        for name, field in klass.__fields__.items():
        # some code to append klass.__name__ and name to a list #
        ...
```
But I personally think the current syntax is a little more intuitive, shorter, and involves less "unwrapping" of objects (I wrote it basically independently of `generate_context()`) - but this could just be my inexperience with `dir()` / `getattr()`.

As to why re-fetch the models + fields at all:
- I thought that this was the intention for the test (i.e., checking that everything in `models.py` ends up in @context,
  _https://github.com/neurobagel/bagelbids/pull/67#discussion_r1107866940_). Not sure if I misunderstood what you had in mind (?), but I personally feel this is a useful sanity check as I find it's not immediately obvious in `generate_context()`, given the conditionals involved, what actually ends up in the context.

Why not use a hardcoded list of the models/fields from models.py (e.g. `["Bagel", "Acquisition", "hasContrastType" ...]`) for the test?
- This would be clearer/simpler, but I think the current method is most robust to things changing in `models.py` (since `generate_context()` uses that file).

Sorry for the word dump! I wasn't sure exactly was the confusion lay, so thought I'd explain my whole thought process. Let me know your thoughts!

Reply:
Thanks a lot for laying it all out so clearly! I agree with all of the points except for the rationale of the last one: 

> This would be clearer/simpler, but I think the current method is most robust to things changing in models.py

You're right that this would probably make the test more robust to us adding / removing / changing things in the `models` file. But that's not the main objective here. The most important thing for a test is to be readable, because it is a form of documentation and because when it fails, you want to easily get a sense of what inputs, expected outputs, and tested behaviour look like. It is there to help you quickly get a sense of what may have caused a problem if there is a problem / the test is failing. So in tests (unlike regular code) it is perfectly OK to have repeated code sections and hard-coded inputs if it aids readability. 

As you said yourself, the `gen_context` function needs a couple of reads to fully understand. That's probably fine for regular code. But it's not great as input to a test where my focus is to get a clear, explicit idea of what inputs and outputs and behaviour are, ideally without having to jump around different sections of the code or having to first understand some other piece of code.

Another issue here is that although your function is a little different from `gen_context` (and possibly more readable), it is still so close to `gen_context` that the test becomes a bit circular: if there is some bug that makes it so a class or attribute in `models.py` doesn't get into the context when we expect it to be there, there's a good chance we won't catch it here if your function and `gen_context` both treat it the same. Conversely, if there is a bug in your function but not in `get_context`, the test would give us a false negative. 

So my suggestion would be to just hard-code the inputs here because we know what the models and attributes and their names are - maybe as a nested list, organized by model. If we inadvertently  remove or rename one, the test should at fail to tell us "something isn't the way I expect it to be". That's a good thing. If we intentionally remove or rename something, we should change the test first anyway, see that it fails, and then make the change in `models.py`. 

If you want, you could run `gen_context` once in a fixture and then pass it to this test as a second input.

Reply:
Yeah, you've convinced me on the circularity and hard-coding! I think you're right that passing in a list is probably the most test-friendly.

I like the nested list idea, but I'm not sure it would be most efficient (?) since some attributes are repeated in multiple models (e.g., `identifier`) but only appear once in the @context - and therefore only need to be tested once, if I'm understanding things correctly. I'd also still like to have the attributes parameterized if possible, for which I think a nested list _may_ get a little wonky. I'll try playing around with it a bit on my end to see what works best with a hardcoded list. 🙂

Reply:
Yeah, you're right, there will be some overlap in attribute names across models. I'm not too concerned at the moment about how fast our tests run - I think if you run `gen_context` only once the individual checks should be pretty fast! But if you notice that the tests get quite slow, we'll have to look into profiling / optimizing them then.

Reply:
Oh, yeah, definitely do the parameterization, that is great! Not sure what the best way is there, but you could probably do something like:

```python
@pytest.mark.parametrize("model, attributes", [
  ("Model", ["attr1", "attr2"]),
  ...
? But really whatever seems the most clear to you is good!

Reply:
Yeah, ended up going with this! It's not the _most_ atomized solution, since as soon as one attribute in `attributes` is not found that test halts, but with the attributes iterated over in a loop, at least we get an informative FAIL message for the test about what was missing. Think this is the best option for now :)

Reply:
also: fixturizing (?) `generate_context` worked like a charm for speeding things up!